pkg/steps/index_generator.go

```diff
@@ -0,0 +1,128 @@
+package steps
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	buildapi "github.com/openshift/api/build/v1"
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/results"
+	imageclientset "github.com/openshift/client-go/image/clientset/versioned/typed/image/v1"
+	coreapi "k8s.io/api/core/v1"
+)
+
+type indexGeneratorStep struct {
+	config      api.IndexGeneratorStepConfiguration
+	resources   api.ResourceConfiguration
+	buildClient BuildClient
+	imageClient imageclientset.ImageStreamsGetter
+	istClient   imageclientset.ImageStreamTagsGetter
+	jobSpec     *api.JobSpec
+	artifactDir string
+	pullSecret  *coreapi.Secret
+}
+
+const IndexOutputDir = "/index-data"
+const IndexDockerfileName = "index.Dockerfile"
+
+func (s *indexGeneratorStep) Inputs() (api.InputDefinition, error) {
+	return nil, nil
+}
+
+func (s *indexGeneratorStep) Run(ctx context.Context) error {
+	return results.ForReason("building_index_generator").ForError(s.run(ctx))
+}
+
+func (s *indexGeneratorStep) run(ctx context.Context) error {
+	source := fmt.Sprintf("%s:%s", api.PipelineImageStream, api.PipelineImageStreamTagReferenceSource)
+	workingDir, err := getWorkingDir(s.istClient, source, s.jobSpec.Namespace())
+	if err != nil {
+		return fmt.Errorf("failed to get workingDir: %w", err)
+	}
+	dockerfile, err := s.indexGenDockerfile()
+	if err != nil {
+		return err
+	}
+	build := buildFromSource(
+		s.jobSpec, api.PipelineImageStreamTagReferenceSource, s.config.To,
+		buildapi.BuildSource{
+			Type:       buildapi.BuildSourceDockerfile,
+			Dockerfile: &dockerfile,
+			Images: []buildapi.ImageSource{
+				{
+					From: coreapi.ObjectReference{
+						Kind: "ImageStreamTag",
+						Name: source,
+					},
+					Paths: []buildapi.ImageSourcePath{{
+						SourcePath:     fmt.Sprintf("%s/.", workingDir),
+						DestinationDir: ".",
+					}},
+				},
+			},
+		},
+		"",
+		s.resources,
+		s.pullSecret,
+	)
+	return handleBuild(ctx, s.buildClient, build, s.artifactDir)
+}
+
+func (s *indexGeneratorStep) indexGenDockerfile() (string, error) {
+	var dockerCommands []string
+	dockerCommands = append(dockerCommands, "")
+	dockerCommands = append(dockerCommands, "FROM opm-builder")
+	dockerCommands = append(dockerCommands, fmt.Sprintf("WORKDIR %s", IndexOutputDir))
+	var bundles []string
+	for _, bundleName := range s.config.OperatorIndex {
+		fullSpec, err := getFullPullSpec(bundleName, s.jobSpec.Namespace(), s.imageClient)
+		if err != nil {
+			return "", fmt.Errorf("failed to get replacement imagestream for image tag `%s`", bundleName)
+		}
+		bundles = append(bundles, fullSpec)
+	}
+	dockerCommands = append(dockerCommands, fmt.Sprintf(`RUN ["opm", "index", "add", "--bundles", "%s", "-d", "%s", "--generate"]`, strings.Join(bundles, ","), IndexDockerfileName))
```

Comment:
nit: is it possible to use a long-form flag, not `-d`? It will be clearer to read and understand for people not intimate with `opm`